gc/structs/ForwardedHeader.hpp

```diff
@@ -243,34 +327,36 @@ class MM_ForwardedHeader
 	MMINLINE bool
 	isSelfForwardedPointer()
 	{
-		return _selfForwardedTag == (_preserved.slot & _selfForwardedTag);
+		return _selfForwardedTag == (getPreservedClassAndTags() & _selfForwardedTag);
 	}
 	
 	MMINLINE bool
 	isStrictlyForwardedPointer()
 	{
 		/* only _forwardedTag set ('self forwarded bit' reset) */
-		return _forwardedTag == (_preserved.slot & _selfForwardedTag);
+		return _forwardedTag == (getPreservedClassAndTags() & _selfForwardedTag);
 	}
 	
 	omrobjectptr_t setSelfForwardedObject();
 	
 	void restoreSelfForwardedPointer();
 	
 	/**
-	 * Initial step for destination object fixup - restore object flags and overlap, while still maintaining progess info and being copied bit.
+	 * Initial step for destination object fixup - restore object flags and overlap, while still maintaining progress info and being copied bit.
 	 */
 	MMINLINE void
 	commenceFixup(omrobjectptr_t destinationObjectPtr)
 	{
-		MutableHeaderFields* newHeader = (MutableHeaderFields *)((fomrobject_t *)destinationObjectPtr + _forwardingSlotOffset);
-		
-		/* copy preserved flags, and keep the rest (which should be all 0s) */
-		newHeader->slot = (_preserved.slot & ~_copyProgressInfoMask) | _beingCopiedTag;
-
-#if defined (OMR_GC_COMPRESSED_POINTERS)
-		newHeader->overlap = _preserved.overlap;
-#endif /* defined (OMR_GC_COMPRESSED_POINTERS) */
+		uintptr_t mask = ~_copyProgressInfoMask;
+#if defined(OMR_GC_COMPRESSED_REFERENCES)
+		if (compressObjectReferences()) {
+			/* _copyProgressInfoMask has the high 32 bits set, so they will be 0 in mask.
+			 * Update mask to not remove the overlap.
+			 */
+			mask |= 0xFFFFFFFF00000000;
+		}
```

Comment:
In theory, this mask should be calculated once, at VM startup. However, I don't see perf issue, since this is not a hot path, so I'm good as is.

Same applies for referenceSize() helper. Could be a pre-calculated variable, but again, not on a hot path.

Reply:
Agree with all these points, and we may address it once the initial work is complete. I believe in the current single-mode builds, these things are still effectively constants.